File: Algoritmos_orden.py
import random
import time

def bubble_sort(arr):
    n = len(arr)
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1): # los últimos i ya están ordenados
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        if not swapped: # salida temprana si ya está ordenado
            break


def selection_sort(arr):
    for i in range(len(arr) - 1):
        min_idx = i                # asume que el mínimo está en i
        for j in range(i + 1, len(arr)):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]


def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]            # elemento a insertar
        j = i - 1
        while j >= 0 and arr[j] > key:     # desplaza mayores hacia la derecha
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key          # inserta key en el hueco


def shell_sort(arr):
    n = len(arr)
    gap = n // 2
    while gap > 0:
        for i in range(gap, n):
            temp = arr[i]      # elemento a insertar
            j = i
            while j >= gap and arr[j - gap] > temp:  # desplaza de a 'gap'
                arr[j] = arr[j - gap]
                j -= gap
            arr[j] = temp         # coloca cuando condición es falsa
        gap //= 2                # reduce el gap a la mitad


def test(data, func, name):
    arr = data.copy()    # copia para no alterar el original
    start = time.time()
    func(arr)
    end = time.time()
    print(f"{name} took {(end - start)*1e6:.2f} μs")


# MAIN — prueba los 4 algoritmos con n = 10, 100, 1000
for n in [10, 100, 1000]:
    data = [random.randint(0, 10000) for _ in range(n)]
    print(f"--- Size: {n} ---")
    test(data, bubble_sort,    "Bubble Sort   ")
    test(data, selection_sort, "Selection Sort")
    test(data, insertion_sort, "Insertion Sort")
    test(data, shell_sort,     "Shell Sort    ")
    print()

File: Algoritmos_orden.cpp
#include <iostream>
#include <vector>
#include <chrono>
using namespace std;
using Clock = chrono::high_resolution_clock;

void bubbleSort(vector<int>& arr) {
    int n = arr.size();
    bool swapped;

    for (int i = 0; i < n - 1; i++) {
        swapped = false;

        // Cada pasada reduce en 1 el rango (los últimos i ya están ordenados)
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
                swapped = true;
            }
        }

        if (!swapped) break; // arreglo ya ordenado, salir temprano
    }
}


void selectionSort(vector<int>& arr) {
    int n = arr.size();

    for (int i = 0; i < n - 1; i++) {
        int minIdx = i; // asume que el mínimo está en i

        // Busca si hay algo menor en el resto del arreglo
        for (int j = i + 1; j < n; j++)
            if (arr[j] < arr[minIdx]) minIdx = j;

        swap(arr[i], arr[minIdx]); // coloca el mínimo en su lugar
    }
}


void insertionSort(vector<int>& arr) {
    int n = arr.size();

    for (int i = 1; i < n; i++) {
        int key = arr[i]; // elemento a insertar
        int j = i - 1;

        // Desplaza hacia la derecha mientras sean mayores que key
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }

        arr[j + 1] = key; // inserta key en el hueco que quedó
    }
}


void shellSort(vector<int>& arr) {
    int n = arr.size();

    // gap: 4 → 2 → 1 (para n=8)
    for (int gap = n / 2; gap > 0; gap /= 2) {

        // Insertion Sort pero comparando de a "gap" posiciones
        for (int i = gap; i < n; i++) {
            int temp = arr[i]; // elemento a insertar
            int j;

            // Desplaza hacia adelante mientras haya mayores
            for (j = i; j >= gap && arr[j - gap] > temp; j -= gap)
                arr[j] = arr[j - gap];

            arr[j] = temp; // inserta cuando la condición es falsa
        }
    }
}


void test(vector<int> data, void(*func)(vector<int>&), const string& name) {
    auto arr = data;              // copia para no alterar el original

    auto t1 = Clock::now();       // marca tiempo de inicio
    func(arr);                    // ejecuta el algoritmo
    auto t2 = Clock::now();       // marca tiempo de fin

    // Imprime nombre y duración en microsegundos
    cout << name << " took "
         << chrono::duration_cast<chrono::microseconds>(t2 - t1).count()
         << " μs\n";
}


int main() {
    vector<int> sizes = {10, 100, 1000};

    for (int n : sizes) {
        // Genera arreglo aleatorio de tamaño n
        vector<int> data(n);
        for (int i = 0; i < n; i++) data[i] = rand();

        cout << "--- Size: " << n << " ---\n";

        // Pasa el mismo arreglo desordenado a cada algoritmo
        test(data, bubbleSort,     "Bubble Sort   ");
        test(data, selectionSort,  "Selection Sort");
        test(data, insertionSort,  "Insertion Sort");
        test(data, shellSort,      "Shell Sort    ");

        cout << "\n";
    }

    return 0;
}